hpx/parallel/algorithms/destroy.hpp

```diff
@@ -0,0 +1,289 @@
+//  Copyright (c) 2014-2017 Hartmut Kaiser
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+/// \file parallel/algorithms/destroy.hpp
+
+#if !defined(HPX_PARALLEL_DETAIL_destroy_JUN_01_2017_1049AM)
+#define HPX_PARALLEL_DETAIL_destroy_JUN_01_2017_1049AM
+
+#include <hpx/config.hpp>
+#include <hpx/traits/is_iterator.hpp>
+#include <hpx/util/void_guard.hpp>
+
+#include <hpx/parallel/algorithms/detail/dispatch.hpp>
+#include <hpx/parallel/algorithms/detail/is_negative.hpp>
+#include <hpx/parallel/execution_policy.hpp>
+#include <hpx/parallel/util/detail/algorithm_result.hpp>
+#include <hpx/parallel/util/loop.hpp>
+#include <hpx/parallel/util/foreach_partitioner.hpp>
+#include <hpx/parallel/util/projection_identity.hpp>
+
+#include <algorithm>
+#include <cstddef>
+#include <iterator>
+#include <memory>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace hpx { namespace parallel { inline namespace v1
+{
+    ///////////////////////////////////////////////////////////////////////////
+    // destroy
+    namespace detail
+    {
+        /// \cond NOINTERNAL
+
+        // provide our own implementation of std::destroy
+        // as some versions of MSVC horribly fail at compiling it for some types
+        // T
+        template <typename FwdIter>
+        void std_destroy(FwdIter first, FwdIter last)
+        {
+            typedef typename std::iterator_traits<FwdIter>::value_type
+                value_type;
+
+            for (/* */; first != last; ++first)
+            {
+                std::addressof(*first)->~value_type();
```

Comment:
Because the type returned from dereferencing `first` could have an overloaded `operator&()`. Alternatively we could write `(*first).~value_type()`, however MSVC was complaining about this for some unknown reason.